File: python/validate_binary_search_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # optimal (preorder traversal)
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isValidSubtree(root, float("-inf"), float("inf"))

    def isValidSubtree(self, root: Optional[TreeNode], lowerBound: int, upperBound: int) -> bool:
        if not root:
            return True

        # root is out of range of parent nodes' values
        if root.val <= lowerBound or root.val >= upperBound:
            return False

        return self.isValidSubtree(root.left, lowerBound, root.val) and self.isValidSubtree(root.right, root.val, upperBound)

File: cpp/validate_binary_search_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // optimal recursive approach
    bool isValidBST(TreeNode *root) {
        return this->isValidSubtree(root, LONG_MIN, LONG_MAX);
    }

private:
    bool isValidSubtree(TreeNode *root, long lowerBound, long upperBound) {
        if (root == NULL) {
            return true;
        }
        // subtree is valid if root within range, left subtree falls within
        // tighter upper bound and right subtree should fall within tighter
        // lower bound
        return lowerBound < root->val && root->val < upperBound &&
               this->isValidSubtree(root->left, lowerBound, root->val) &&
               this->isValidSubtree(root->right, root->val, upperBound);
    }
};
